File: app/src/main/java/com/example/myapplication/core/program/ProgramPhasePlanner.kt
package com.example.myapplication.core.program

enum class ProgramPhase {
    FOUNDATION,
    BUILD,
    CONSOLIDATE,
    DELOAD,
}

object ProgramPhasePlanner {
    fun phaseFor(week: Int, durationWeeks: Int): ProgramPhase {
        require(durationWeeks > 0) { "Program duration must be positive." }
        require(week in 1..durationWeeks) { "Week must be within the program duration." }
        if (durationWeeks >= 4 && week == durationWeeks) return ProgramPhase.DELOAD
        val progress = week.toDouble() / durationWeeks
        return when {
            progress <= 0.25 -> ProgramPhase.FOUNDATION
            progress <= 0.70 -> ProgramPhase.BUILD
            else -> ProgramPhase.CONSOLIDATE
        }
    }
}

File: flutter/lib/core/program/program_phase_planner.dart
enum ProgramPhase {
  foundation,
  build,
  consolidate,
  deload,
}

class ProgramPhasePlanner {
  static ProgramPhase phaseFor(int week, int durationWeeks) {
    if (durationWeeks <= 0) {
      throw ArgumentError("Program duration must be positive.");
    }
    if (week < 1 || week > durationWeeks) {
      throw ArgumentError("Week must be within the program duration.");
    }
    
    if (durationWeeks >= 4 && week == durationWeeks) {
      return ProgramPhase.deload;
    }
    
    final progress = week / durationWeeks.toDouble();
    if (progress <= 0.25) {
      return ProgramPhase.foundation;
    } else if (progress <= 0.70) {
      return ProgramPhase.build;
    } else {
      return ProgramPhase.consolidate;
    }
  }
}
